### PR TITLE
Feature: Add automated testing for notebooks and widgets

### PR DESCRIPTION
**Pull request by jannesgg**
This PR adds automated testing in the form of a testing script which is run on the dev branch when changes are pushed. Builds are also conditional on the tests passing.

### CODE REVIEW

File: test/notebook-tests.py [hunk 1]
```diff
@@ -63,6 +46,8 @@
 
 
 def test_t1():
+    import kso_utils.widgets as kso_widgets
```
Comment (Diewertje11) [inline test/notebook-tests.py:48]: Why did you move all the imports to inside the test functions and not just at the top of the document?

File: test/notebook-tests.py [hunk 2]
```diff
@@ -201,93 +200,123 @@ def test_t5():
...
+    mlp.detect_yolov5(
+        exp_name=exp_name,
+        source=source,
+        save_dir=save_dir,
+        conf_thres=conf_thres.value,
+        artifact_dir=artifact_dir,
+    )
+    eval_dir = exp_path
+    mlp.save_detections_wandb(conf_thres.value, model, eval_dir)
+
```
Comment (Diewertje11) [inline test/notebook-tests.py:259]: Maybe put a comment that the 'Investigating training and validation dataset' is not tested, and that we now go to track unique individuals.

File: test/notebook-tests.py [hunk 3]
```diff
@@ -201,93 +200,123 @@ def test_t5():
...
+    pp.get_zoo_info(zoo_cred=[zoo_user, zoo_pass])
+    pp.choose_workflows(generate_export=False, zoo_cred=[zoo_user, zoo_pass])
+
+    workflow_checks = {
+        "Workflow name: #0": "Development workflow",
+        "Subject type: #0": "clip",
+        "Minimum workflow version: #0": 1.0,
+    }
+
+    class_df = pp.get_classifications(
```
Comment (Diewertje11) [inline test/notebook-tests.py:308]: I do not see this one in the dev version of NB 8. Maybe add a comment where it is?

Because in all the other tests, it is quite easy to follow the test function with the notebook next to it. So that is great! But here someone could get a bit confused since it is not in the NB (or I just don't see it).

**Review by Diewertje11 — COMMENTED** [review]
Looks great! So nice that we have this in place now. I added a few comments with things that are unclear to me. But the rest looks good and it is easy to follow with the notebook next to it. So if we ever change things in the notebook, it will be clear where to change this in the tests as well.

The only thing I had a comment on is for a the kso_utils and a few parts that have the if test=True there. But I am not sure if there is a good solution for it. And I have some comments with questions. (I think you can see the comments in the kso_utils auto-tests by just going into all your commits separately. (For a next time, do you know if there is a way I could include them in this review?)

And I have not run the code to see if it works, but I assume that you have already done that. So I only looked if all code made sense to me.

File: kso_utils
[the remark this answers is not in the document]
Comment (Diewertje11) [reply kso_utils:1]: And some other comments in the other commits in kso-utils as well

File: test/notebook-tests.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+    pp.update_meta(sites_sheet_df, "sites", test=True)
+
+    # --- Automatic check of movies metadata
+    review_method = kso_widgets.choose_movie_review()
+    gpu_available = kso_widgets.gpu_select()
+    pp.check_movies_meta(
+        review_method=review_method.value, gpu_available=gpu_available.result
+    )
+    assert len(sites_sheet_df) == 5
```
[the remark this answers is not in the document]
Comment (jannesgg) [reply test/notebook-tests.py:85]: Exactly, this is the length of the default sites.csv. This is a sanity check on the length after the metadata is loaded in.

File: test/notebook-tests.py (hunk 1 above)
Comment (jannesgg) [reply test/notebook-tests.py:48]: I find it useful to have the imports listed that are in used in each test. That way we can easily see what is needed. If there are functions which are common to all tests, we could move them out.
